File: src/Functional/Haskell/Example1.hs
class Button a where
  renderButton :: a -> IO ()

data DarkButton = DarkButton
data LightButton = LightButton

instance Button DarkButton where
  renderButton _ = putStrLn "Dark Button"

instance Button LightButton where
  renderButton _ = putStrLn "Light Button"

class Checkbox a where
  renderCheckbox :: a -> IO ()

data DarkCheckbox = DarkCheckbox
data LightCheckbox = LightCheckbox

instance Checkbox DarkCheckbox where
  renderCheckbox _ = putStrLn "Dark Checkbox"

instance Checkbox LightCheckbox where
  renderCheckbox _ = putStrLn "Light Checkbox"

data UIFactory = UIFactory
  { createButton :: IO (),
    createCheckbox :: IO ()
  }

darkFactory :: UIFactory
darkFactory = UIFactory
  { createButton = renderButton DarkButton,
    createCheckbox = renderCheckbox DarkCheckbox
  }

lightFactory :: UIFactory
lightFactory = UIFactory
  { createButton = renderButton LightButton,
    createCheckbox = renderCheckbox LightCheckbox
  }

createUIComponents :: UIFactory -> IO ()
createUIComponents factory = do
  createButton factory
  createCheckbox factory

main :: IO ()
main = do
  createUIComponents darkFactory
  createUIComponents lightFactory

File: src/Functional/Scala/Example1.scala
trait Button {
  def render(): Unit
}

class DarkButton extends Button {
  def render(): Unit = println("Dark Button")
}

class LightButton extends Button {
  def render(): Unit = println("Light Button")
}

trait Checkbox {
  def render(): Unit
}

class DarkCheckbox extends Checkbox {
  def render(): Unit = println("Dark Checkbox")
}

class LightCheckbox extends Checkbox {
  def render(): Unit = println("Light Checkbox")
}

trait UIFactory {
  def createButton(): Button
  def createCheckbox(): Checkbox
}

class DarkFactory extends UIFactory {
  def createButton(): Button = new DarkButton
  def createCheckbox(): Checkbox = new DarkCheckbox
}

class LightFactory extends UIFactory {
  def createButton(): Button = new LightButton
  def createCheckbox(): Checkbox = new LightCheckbox
}

object Example1 {
  def main(args: Array[String]): Unit = {
    val darkFactory: UIFactory = new DarkFactory
    val lightFactory: UIFactory = new LightFactory

    val darkButton = darkFactory.createButton()
    val darkCheckbox = darkFactory.createCheckbox()
    darkButton.render()
    darkCheckbox.render()

    val lightButton = lightFactory.createButton()
    val lightCheckbox = lightFactory.createCheckbox()
    lightButton.render()
    lightCheckbox.render()
  }
}
